content/blog/2020/07/2020-07-27-machine-learning-plugin-coding-phase2.adoc

```diff
@@ -0,0 +1,150 @@
+[source]
+----
+FROM jenkins/agent:latest
+
+MAINTAINER Loghi <loghijiaha@gmail.com>
+
+USER root
+
+RUN apt update && apt install --no-install-recommends python3 -y \
+    python3-pip \
+    && rm -rf /var/lib/apt/lists/*
+
+COPY requirements.txt /requirements.txt
+
+RUN pip3 install --upgrade pip setuptools && \
+    pip3 install --no-cache-dir -r /requirements.txt && \
+    ln -sf /usr/bin/python3 /usr/bin/python && \
+    ln -sf /usr/bin/pip3 /usr/bin/pip
+
+USER jenkins
+----
+
+
+=== Ported to Python 3
+
+As discussed in the previous meeting, we concluded that the plugin should support Python 3 as Python 2.7+ has been deprecated since the beginning of 2020. Pull request for docker agent should be also ported to Python 3 support.
+
+=== Jupyter Rest Client API
+
+The Jupyter Notebook server API seemed to be promising that it can be also used to run notebooks and codes. There were 3 api implementations that were merged in the master. But we had to focus on what was proposed in the design document and had to finish all must-have issues/works. Jupyter REST client was left for future implementation. It is also a good start to contribute to the plugin from the community.
+
+
+=== Fixed bugs for running in agent
+
+There were a few bugs related to the file path of notebooks while building a job. The major problem was caused by the python dependencies needed to connect to a IPython kernel. All issues/bugs were fixed before the timeline given.
+
+=== R support as a future improvement
+
+This is what we tried to give a glimpse of knowledge that this plugin can be extended for multi language support in the future. There was a conclusion that the kernel should be selected dynamically using extension of the script file(like `eval_model.rb` or `train_model.r`), instead of scripting the same code for each kernel.
+
+=== Documentation and End to End testing
+
+A well explained documentation was published in the repository. A guided tutorial to run a notebook checked out from a git repo in an agent was included in the link:https://github.com/jenkinsci/machine-learning-plugin/blob/master/docs/ML-docker-agent.adoc[docs] page. Mentors helped to test our plugin in both Linux and Windows.
+
+
+=== Code editor with rebuild feature
+
+Code editor was filtered as a nice to have feature in the design document. After grabbing the idea of Jenkinsfile replay editor, I could do the same for the code. At the same time, when we are getting the source code from git, it is not an elegant way of editing code in the original code. After the discussion, we had to leave the PR open that may have use cases in the future if needed.
+
+=== Jenkins LTS update
+
+The plugin has been updated to support Jenkins LTS 2.204.1 as 2.164.3 had some problems with installing pipeline supported API/plugin
+
+== Installation for experimental version
+
+. Enable the link:/doc/developer/publishing/releasing-experimental-updates/#using-the-experimental-update-center[experimental update center]
```

Comment:
Maybe it makes sense to cut a release candidate (e.g. version `1.0-rc-1`) so that the plugin appears in the main update centers and on the plugin site. It is not required for the blogpost, but something for a discussion @Loghijiaha @kinow

Reply:
If that's necessary I'd be fine with that. Only concern I have is if this gets more attention now, @Loghijiaha could have to spend more time helping/answering user questions (which already happened twice so far I think, even though it's in the experimental). Maybe when we get closer to the end of the GSoC?

There are also updates that are breaking backward compatibility, so `rc-1` and `rc-2` would not be fully compatible (which I think some users would expect, but we could also make it explicit in the plugin readme or description that appears in the update center)

Reply:
> If that's necessary I'd be fine with that. 

It is not necessary, up to the project team to decide